extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+      }
+    }
+
+    return retVal;
+  }
+
+  private void contactWorkersForStage(final TaskContactFn contactFn, final IntSet workers)
+  {
+    final List<String> taskIds = getTaskIds();
+    final List<ListenableFuture<Void>> taskFutures = new ArrayList<>(workers.size());
+
+    for (int workerNumber : workers) {
+      final String taskId = taskIds.get(workerNumber);
+      taskFutures.add(contactFn.contactTask(netClient, taskId, workerNumber));
+    }
+
+    FutureUtils.getUnchecked(MSQFutureUtils.allAsList(taskFutures, true), true);
+  }
+
+  private void startWorkForStage(
+      final QueryDefinition queryDef,
+      final ControllerQueryKernel queryKernel,
+      final int stageNumber,
+      @Nullable final List<SegmentIdWithShardSpec> segmentsToGenerate
+  )
+  {
+    final Int2ObjectMap<Object> extraInfos = makeWorkerFactoryInfosForStage(
+        queryDef,
+        stageNumber,
+        queryKernel.getWorkerInputsForStage(queryKernel.getStageId(stageNumber)),
+        segmentsToGenerate
+    );
+
+    final Int2ObjectMap<WorkOrder> workOrders = queryKernel.createWorkOrders(stageNumber, extraInfos);
+
+    contactWorkersForStage(
+        (netClient, taskId, workerNumber) -> netClient.postWorkOrder(taskId, workOrders.get(workerNumber)),
+        workOrders.keySet()
+    );
+  }
+
+  private void postResultPartitionBoundariesForStage(
+      final QueryDefinition queryDef,
+      final int stageNumber,
+      final ClusterByPartitions resultPartitionBoundaries,
+      final IntSet workers
+  )
+  {
+    contactWorkersForStage(
+        (netClient, taskId, workerNumber) ->
+            netClient.postResultPartitionBoundaries(
+                taskId,
+                new StageId(queryDef.getQueryId(), stageNumber),
+                resultPartitionBoundaries
+            ),
+        workers
+    );
+  }
+
+  /**
+   * Publish the list of segments. Additionally, if {@link DataSourceMSQDestination#isReplaceTimeChunks()},
+   * also drop all other segments within the replacement intervals.
+   *
+   * If any existing segments cannot be dropped because their intervals are not wholly contained within the
+   * replacement parameter, throws a {@link MSQException} with {@link InsertCannotReplaceExistingSegmentFault}.
+   */
+  private void publishAllSegments(final Set<DataSegment> segments) throws IOException
+  {
+    final DataSourceMSQDestination destination =
+        (DataSourceMSQDestination) task.getQuerySpec().getDestination();
+    final Set<DataSegment> segmentsToDrop;
+
+    if (destination.isReplaceTimeChunks()) {
+      final List<Interval> intervalsToDrop = findIntervalsToDrop(Preconditions.checkNotNull(segments, "segments"));
+
+      if (intervalsToDrop.isEmpty()) {
+        segmentsToDrop = null;
+      } else {
+        segmentsToDrop =
+            ImmutableSet.copyOf(
+                context.taskActionClient().submit(
+                    new RetrieveUsedSegmentsAction(
```

Comment:
Race condition. If I undersand this logic correctly, we ask Overlord for a list of segments, do some computation, and then request to publish or remove segments. This work is done in this process, outside of any locking.

It should be Overlord (within a lock) that does this work. Else, the information can be stale here by the time we do the computations and send the replace/add actions.

Reply:
This scenario is OK, because this block is happening only for REPLACE (`destination.isReplaceTimeChunks()` is `true`), and in those cases the task acquires locks for the lifetime of the task (see `MSQControllerTask.isReady`). I'll add some comments spelling this out.